contracts/Orders.sol

```diff
@@ -0,0 +1,235 @@
+    )
+        private
+        returns(uint32)
+    {
+        Order storage prevOrder = orders[prevId];
+
+        // Add new order
+        uint32 orderId = nextId++;
+        orders[orderId] = Order({
+            maker: maker,
+            srcAmount: srcAmount,
+            dstAmount: dstAmount,
+            prevId: prevId,
+            nextId: prevOrder.nextId
+        });
+
+        // Update next order to point back to added order
+        uint32 nextOrderId = prevOrder.nextId;
+        if (nextOrderId != TAIL_ID) {
+            Order storage nextOrder = orders[nextOrderId];
+            nextOrder.prevId = orderId;
+        }
+
+        // Update previous order to point to added order
+        prevOrder.nextId = orderId;
+
+        return orderId;
+    }
+
+    function getOrderDetails(uint32 orderId)
+        internal
+        view
+        returns (
+            address _maker,
+            uint128 _srcAmount,
+            uint128 _dstAmount,
+            uint32 _prevId,
+            uint32 _nextId
+        )
+    {
+        Order storage order = orders[orderId];
+        return (
+            order.maker,
+            order.srcAmount,
+            order.dstAmount,
+            order.prevId,
+            order.nextId
+        );
+    }
+
+    function removeById(uint32 orderId) internal {
+        verifyCanRemoveOrderById(orderId);
+
+        // Disconnect order from list
+        Order storage order = orders[orderId];
+        orders[order.prevId].nextId = order.nextId;
+        orders[order.nextId].prevId = order.prevId;
+
+        order.prevId = 0;
+        order.nextId = 0;
+    }
+
+    // The updated order id is returned following the update.
+    function update(uint32 orderId, uint128 srcAmount, uint128 dstAmount)
+        internal
+        returns(uint32)
+    {
+        address maker = orders[orderId].maker;
+        removeById(orderId);
+        return add(maker, srcAmount, dstAmount);
+    }
+
+    // The updated order id is returned following the update.
+    function updateWithPositionHint(
+        uint32 orderId,
+        uint128 srcAmount,
+        uint128 dstAmount,
+        uint32 prevId
+    )
+        internal
+        returns(uint32)
+    {
+        address maker = orders[orderId].maker;
+        removeById(orderId);
+        return addAfterId(maker, srcAmount, dstAmount, prevId);
+    }
+
+    function allocateIds(uint32 howMany) internal returns(uint32) {
+        uint32 firstId = nextId;
+        nextId += howMany;
+        return firstId;
+    }
+
+    function verifyCanRemoveOrderById(uint32 orderId) private view {
+        require(orderId != BUY_HEAD_ID);
+
+        Order storage order = orders[orderId];
+
+        // Make sure such order exists in mapping.
+        require(order.prevId != 0 || order.nextId != 0);
+    }
+
+    function calculateOrderSortKey(uint128 srcAmount, uint128 dstAmount)
+        public
+        pure
+        returns(uint)
+    {
+        return dstAmount * PRECISION / srcAmount;
```

Comment:
check over flow....
or use safe math???
better just check. safe math is annoying.

Reply:
Let's discuss this.